### PR TITLE
TASK: minor clean up fusion runtime code 

### PR DESCRIPTION
**Pull request by mhsdesign**
I often wanted to bring some light into the Runtime, now here it is ^^

I made **really** small, descriptive commits, so it should be superclear why i did touch code part x.

I ran the benchmark afterwards but there is no difference, as i only changed the syntax...

### CODE REVIEW

**Comment by kdambekalns** [comment]
> i must have been really bored to change all those `sprintf` to string interpolation out of principle

Just being curious!: What principle would that be?

To me `sprintf()` is way more explicit, easier to read with multiple variables and e.g. PhpStorm warns you in case if mismatches between placeholder and variable count… So far in cleanups we rather went into the other direction, adding `sprintf` instead of interpolation.

**Comment by mhsdesign** [comment]
i think that would be my own taste, which i developed... ;)

i think neither psr2 or psr12 says whats the way to go ... but im happy to revert these things and stick in the future to sprintf, if that is what neos agrees upon.

i personally dont like the syntax. String manipulition looks like an actual template (and im used to it from pythons `f" {bar} "` and javascripts
```
`${bar}`
```

**Comment by fcool** [comment]
Nice discussion... I'm personally still not convinced from one method over the other. (I usually tend to use string concatenation over interpolation)

But I've seen benchmarks, which clearly show, that the PHP string interpolation is the "best" option, when it comes to performance. So that might be an argument.
See https://stackoverflow.com/a/58917133

**Comment by kdambekalns** [comment]
> […] when it comes to performance. So that might be an argument.
> See https://stackoverflow.com/a/58917133

Not really. The difference is only visible with thousands of iterations or more, and thus not relevant. Rather, as pointed out at the very same link:

> Use whatever makes your code most readable and maintainable for you and others

**Comment by fcool** [comment]
Yay! Readibility should be always the main target.

The "minor difference" is never the less an argument, as there is a difference. And even an infinitesimal difference can get big enough to matter, if you iterate only often enough. ;)

**Comment by mhsdesign** [comment]
I agree that we can outrule any performance implications, as this is mostly concerning exception messages

**Comment by fcool** [comment]
Just for the record: I am not saying performance is a point to consider HERE, but understood Karsten's question as "general question".
So we are mixing two streams here: "General code style" and "makes a difference here".

When we come back to readibility and stay with exactly this example:
For me every change in this PR is now way more easier to understand and follow than before. Especially when it comes to formatting and setting parts of names, etc.

**Comment by kitsunet** [comment]
I think we should find one way to agree on to write those messages, whatever it is. I do like this proposal and at least phpstorm correctly recognises the variables used inside the string and would properly refactor them. So this way > string concat. I am undecided about sprintf, I do like the specific enumeration of variables, but backtracing a longer message for which variable is which can be annoying as well. So 50/50 no the two options I guess. Given that the majority of code is sprintf atm I am leaning towards (keeping) that as "standard".

File: Neos.Fusion/Classes/Core/Runtime.php [hunk 1]
```diff
@@ -907,25 +871,22 @@ protected function throwExceptionForUnrenderablePathIfNeeded($fusionPath, $fusio
...
-            ), 1332493995);
+            throw new Exceptions\MissingFusionImplementationException(<<<MESSAGE
+                The Fusion object "$objectType" cannot be rendered:
+                Most likely you mistyped the prototype name or did not define
+                the Fusion prototype with "prototype($objectType) < prototype(...)".
+                Other possible reasons are a missing parent-prototype or
+                a missing "@class" annotation for prototypes without parent.
+                It is also possible your Fusion file is not read because
+                of a missing "include:" statement.
+                MESSAGE, 1332493995);
```
Comment (mhsdesign) [inline Neos.Fusion/Classes/Core/Runtime.php:882]: I like my proposed Heredoc much better for multilines. Also note that there will be no unnecessary whitespace.

```
sprintf(
    "Fooobar
     bung buz"
)
```
will result in `"Fooobar\n    bung buz"`

but



```
<<<MESSAGE
    "Fooobar
     bung buz"
     MESSAGE
```
will result in `"Fooobar\nbung buz"`

**Comment by fcool** [comment]
I would love not to reglement this part of "code style". And Marc-Henry has good points here. To not have weird whitespace in the error messages one would have both: String concatenation and sprintf interpolation.

Is it really an issue big enough to enforce a "standard" way?

**Review by kdambekalns — APPROVED** [review]

**Comment by kdambekalns** [comment]
Let's not block this because of a handful of converted `sprintf` calls… HEREDOC is great for multiline things, anything that needs type-casting or "advanced" formatting is probably better dealt with using `sprintf`.

So far we got along well, so… 🤷‍♂️

**Review by PRGfx — COMMENTED** [review]
I like the changes of using new syntax elements to clean up no longer necessary constructs.

Seeing the discussion on string formatting I agree that it would be nice to have a common style throughout the code base. However the proposed style looks fine to me here.

I'm not a big fan of the method inlining. I recognize these methods not only as a means to deduplicate code, but also to extract code into a descriptive name with the possibility to explain some of the reasoning of what is happening, if necessary (as for `finalizePathEvaluation`). I'm a bit torn whether an inline comment does better - after all it's right where the reader would need it - if the method would not be named appropriately.
`isArrayFusionObject` is a bit of an odd one to me: I prefer a name to some wild `instanceof` call, however I dislike that name. The whole point of the `ArrayFusionObject` is that it gets all properties assigned. That property is not well reflected in the name imho however in an effort to clean things up _here_, I could imagine something more descriptive along the lines of `shouldAssignProperties`.

**Comment by mhsdesign** [comment]
thanks for your review and your opinion @PRGfx 

> I'm not a big fan of the method inlining. I recognize these methods not only as a means to deduplicate code, but also to extract code into a descriptive name with the possibility to explain some of the reasoning of what is happening, if necessary (as for `finalizePathEvaluation`). I'm a bit torn whether an inline comment does better - after all it's right where the reader would need it - if the method would not be named appropriately.

In principle i share your opinion: better add a method than inlining it and adding comments.
But in this specific case I do not agree.

Generally I dont really like methods such as `push` and `pop` as they are closely interlocked - and if one misses to call the later, it screws up. _I much rather prefer to see such state inside a `withContext` method, that takes in a closure which will be evaluated inside a special state (much like pythons context managers)_

But in our case we already have the classical `push` and `pop` see `$runtimeContentCache->enter` and `$runtimeContentCache->leave` (and also a lot of state concerning context and applyValues which needs to be pushed and popped)

Previously the `$runtimeContentCache->enter` was in the method `evaluate` and `$runtimeContentCache->leave` was extracted alongside popping context and applyValues inside `finalizePathEvaluation`.

I dislike this approach as this seems like a code smell, which was only needed because we needed to call `finalizePathEvaluation` multiple times. Due to more modern syntax and precise refactoring we now only need to execute this logic in one place and have the possibility to inline it.

Also `popContext` and `popApplyValues` should happen directly inlined in that place aswell (as i proposed).
I think their naming is already quite descriptive and could not be more legible.

> `isArrayFusionObject` is a bit of an odd one to me: I prefer a name to some wild `instanceof` call, however I dislike that name. The whole point of the `ArrayFusionObject` is that it gets all properties assigned. That property is not well reflected in the name imho however in an effort to clean things up _here_, I could imagine something more descriptive along the lines of `shouldAssignProperties`.

here i agree with your point _**I prefer a name to some wild `instanceof` call**_ definitely convinces me.
i will think of a name...

**Comment by mhsdesign** [comment]
> > `isArrayFusionObject` is a bit of an odd one to me: I prefer a name to some wild `instanceof` call, however I dislike that name. The whole point of the `ArrayFusionObject` is that it gets all properties assigned. That property is not well reflected in the name imho however in an effort to clean things up _here_, I could imagine something more descriptive along the lines of `shouldAssignProperties`.
> 
> here i agree with your point _**I prefer a name to some wild `instanceof` call**_ definitely convinces me. i will think of a name...

i adjusted the naming in https://github.com/neos/neos-development-collection/pull/3674/commits/365f80ad429c86b64e02ce35a61745b0b1323e6d
